### PR TITLE
Support advanced Node.js ESM import types

### PR DESCRIPTION
**Pull request by yankovalera**
Support advanced Node.js ESM import types. Relates to #57

### CODE REVIEW

**Comment by ryan-roemer** [comment]
Thanks for this! Can I get this PR without the extraneous formatting that was also added so that the diff is easier to review?

**Comment by yankovalera** [comment]
@ryan-roemer I did eslint --fix and this is what I got, any ideas if I'm missing something?

**Comment by ryan-roemer** [comment]
Did you have something else format first? This project doesn’t use prettier but it looks like it was auto applied at some point…

**Comment by yankovalera** [comment]
Looks like VScode applied some formatting - it's looking good now. Do you think it makes sense to add prettier?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,14 +25,14 @@ jobs:
         with:
           node-version: ${{ matrix.node-version }}
 
-      # Caching: root project
+      # Note: Yarn root cache restore is slow (1:30) on Windows, so only do on Linux.
```
Comment (ryan-roemer) [inline .github/workflows/ci.yml:28]: This is unrelated -- but slowing us down on Windows, so just grabbed it here.

File: lib/trace.js [hunk 2]
```diff
@@ -210,6 +212,18 @@ const _resolveDep = async ({
   bailOnMissing,
   addRootPackagePaths = false
 }) => {
+  // Support advanced Node.js ESM import types by ignoring built-in's as not
+  // needing any further inspection / inclusion.
+  if (IGNORE_IMPORT_PREFIXES.find((prefix) => depName.startsWith(prefix))) {
```
Comment (ryan-roemer) [inline lib/trace.js:217]: I switched from `indexOf() !== -1` to `startsWith()` to be a little bit more of a semantic match with the fact that we're only matching prefixes, not string tokens anywhere throughout the target string.

File: lib/trace.js [hunk 3]
```diff
@@ -210,6 +212,18 @@ const _resolveDep = async ({
...
+    return null;
+  }
+
+  // Check for modern import `?` query imports and mutate dep name to remove.
+  const queryStringPosition = depName.indexOf("?");
+  if (queryStringPosition > -1) {
+    depName = depName.substring(0, queryStringPosition);
```
Comment (ryan-roemer) [inline lib/trace.js:224]: Since we use `depName` elsewhere in this function, it seemed safer to just mutate the variable at the top before anything else started using it...

File: test/lib/trace.spec.js [hunk 4]
```diff
@@ -2345,6 +2345,139 @@ describe("lib/trace", () => {
         });
       });
     });
+
+    describe("modern ESM imports", () => {
```
Comment (ryan-roemer) [inline test/lib/trace.spec.js:2349]: I typically do the nitty-gritty feature testing on `traceFile` and leave `traceFiles` tests to things involving multiple `srcPaths`, so I moved the existing tests in this describe block and targeted `traceFile`.

The tests provided great coverage, and I just added a few more corner cases to keep everything covered 😄

**Review by ryan-roemer — APPROVED** [review]
@yankovalera Fantastic work! Everything works great and this will be great to land as more library authors start using these constructs (I've seen chatter especially about folks starting to use `node:` imports).

As I tuned and tweaked a few things, I'm also adding you as a reviewer to do a final pass and approve. After you approve, you can either merge and release (I'm thinking `0.4.5`) following the instructions in `CONTRIBUTING.md` (and I think you'll need to ping me to add you to our npm publishing group) or just ping me and I'll do the release. Cheers!

File: lib/trace.js (hunk 2 above)
Comment (yankovalera) [reply lib/trace.js:217]: Thanks - it's indeed more performant.

File: test/lib/trace.spec.js (hunk 4 above)
Comment (yankovalera) [reply test/lib/trace.spec.js:2349]: thanks!

**Comment by ryan-roemer** [comment]
Released in `trace-deps@0.4.5`
